### DISCUSSION TITLE
Is there any bulk methods for AsnycSession?

### DISCUSSION THREAD

**Discussion by teamhide** [sqlalchemy/sqlalchemy#7869] [Usage Questions]
Previously, I was using sqlalchemy version 1.3, but I found out that it supports asynchronous from version 1.4.

In the case of the previous synchronous session, bulk operations were being performed through methods such as `bulk_insert_mappings` and `bulk_save_objects`.

When I try to use the same in version 1.4, the following error occurs.

```python
AttributeError: 'async_scoped_session' object has no attribute 'bulk_insert_mappings'
```

Currently, I used session that created through `sessionmaker()` and `async_scoped_session()`

```python
engine = create_async_engine(config.WRITER_DB_URL, pool_recycle=3600)
async_session_factory = sessionmaker(
    bind=engine, class_=AsyncSession,
)
session: Union[AsyncSession, async_scoped_session] = async_scoped_session(
    session_factory=async_session_factory,
    scopefunc=get_session_context,
)
```

Even though I searched the official documentation and many times, there was no information related to bulk operation. 

Is bulk operation impossible in asynchronous mode?

**Comment by zzzeek** [comment]
> ```python
> engine = create_async_engine(config.WRITER_DB_URL, pool_recycle=3600)
> async_session_factory = sessionmaker(
>     bind=engine, class_=AsyncSession,
> )
> session: Union[Session, async_scoped_session] = async_scoped_session(
>     session_factory=async_session_factory,
>     scopefunc=get_session_context,
> )
> ```
> 
> Even though I searched the official documentation and many times, there was no information related to bulk operation.
> 
> Is bulk operation impossible in asynchronous mode?

bulk operations are available by using insert() statements directly, which should be preferred to the bulk_save methods which should never have been added and I hope to remove them someday.      For example see the use of an upsert at https://docs.sqlalchemy.org/en/14/orm/persistence_techniques.html#using-postgresql-on-conflict-with-returning-to-return-upserted-orm-objects .    2.0 will continue to refine the use of insert() statements for bulk ORM operations.
